### PR TITLE
Implement autodiscover in case of empty or absent machete file

### PR DESCRIPTION
**Pull request by micpiotrowski**
- [ ] once it's merged remember to update general usage instructions

### CODE REVIEW

File: frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/speedsearch/DiscoverAction.java [hunk 1]
```diff
@@ -1,27 +1,28 @@
-package com.virtuslab.gitmachete.frontend.actions.graphtable;
+package com.virtuslab.gitmachete.frontend.actions.speedsearch;
```
Comment (micpiotrowski) [inline frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/speedsearch/DiscoverAction.java:1]: Changes in this file are unrelated

Comment (mkondratek) [reply frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/speedsearch/DiscoverAction.java:1]: I would rather avoid piggybacking anything on bug fixing PRs (to lower the chances of providing another bug :rofl:)

---

Anyway... Why this change (rename)?
Where is the connection between this package and the IntelliJ speed search?

Afaik the speed search is an ability to simply write a text within some window/dialog and search (results marked with yellow on the light theme). If there is no connection to this feature please revert this change / rename to smth else to avoid confusion.

Comment (PawelLipski) [reply frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/speedsearch/DiscoverAction.java:1]: Yeah, I don't like the `speedsearch` name either... just `unanchored` (?) would be better if this action doesn't belong to any particular `ActionPlace`

File: frontendResourceBundles/src/main/resources/GitMacheteBundle.properties [hunk 2]
```diff
@@ -209,6 +209,16 @@ string.GitMachete.BranchOrCommitCellRendererComponent.sync-to-remote-status-text
 string.GitMachete.BranchOrCommitCellRendererComponent.sync-to-remote-status-text.diverged-from-and-older-than-remote=(diverged from & older than {0})
 string.GitMachete.BranchOrCommitCellRendererComponent.sync-to-remote-status-text.untracked=(untracked)
 
+string.GitMachete.EnhancedGraphTable.automatic-discover.cant-discover-layout-error-title=Can't automatically discover repository layout
+string.GitMachete.EnhancedGraphTable.automatic-discover.task-title=Automatic discovering branch layout
```
Comment (PawelLipski) [inline frontendResourceBundles/src/main/resources/GitMacheteBundle.properties:213]:
```suggestion
string.GitMachete.EnhancedGraphTable.automatic-discover.task-title=Automatically discovering branch layout
```

File: frontendResourceBundles/src/main/resources/GitMacheteBundle.properties [hunk 3]
```diff
@@ -209,6 +209,16 @@ string.GitMachete.BranchOrCommitCellRendererComponent.sync-to-remote-status-text
...
+string.GitMachete.EnhancedGraphTable.automatic-discover.success-message=Branch layout has been automatically detected
```
Comment (PawelLipski) [inline frontendResourceBundles/src/main/resources/GitMacheteBundle.properties:214]: I'd rather use `discover` consistently instead of `detect`

File: frontendUiTableImpl/src/main/java/com/virtuslab/gitmachete/frontend/ui/impl/table/EnhancedGraphTable.java [hunk 4]
```diff
@@ -208,13 +209,74 @@ private void refreshModel(GitRepository gitRepository, List<String> skippedBranc
...
+  private void doAutomaticDiscover(Path macheteFilePath) {
+    var selectedRepository = getGitRepositorySelectionProvider().getSelectedGitRepository().getOrNull();
+    if (selectedRepository == null) {
+      LOG.error("Can't do automatic discover because of undefined selected repository");
+      return;
+    }
+    var mainDirPath = GitVfsUtils.getMainDirectoryPath(selectedRepository).toAbsolutePath();
+    var gitDirPath = GitVfsUtils.getGitDirectoryPath(selectedRepository).toAbsolutePath();
+
+    @SuppressWarnings({"nullness", "guieffect"}) // To allow runResult be uninitialized and Tuple accept @Nullable IBranchLayout
```
Comment (PawelLipski) [inline frontendUiTableImpl/src/main/java/com/virtuslab/gitmachete/frontend/ui/impl/table/EnhancedGraphTable.java:221]: Hmm what's the specific error caused by `runResult` being uninitialized? :thinking: Is it about dereference of `runResult` in line 245?

Comment (PawelLipski) [reply frontendUiTableImpl/src/main/java/com/virtuslab/gitmachete/frontend/ui/impl/table/EnhancedGraphTable.java:221]: If so, then consider checking against `runResult == null` in `onSuccess`

File: frontendUiTableImpl/src/main/java/com/virtuslab/gitmachete/frontend/ui/impl/table/EnhancedGraphTable.java [hunk 5]
```diff
@@ -208,13 +209,74 @@ private void refreshModel(GitRepository gitRepository, List<String> skippedBranc
...
+              return Tuple(branchLayoutWriter, branchLayout, repositorySnapshot);
+            })
+            .filterTry(tuple -> tuple._2 != null, __ -> new EmptyBranchLayoutException())
+            .andThenTry(tuple -> tuple._1.write(macheteFilePath, tuple._2, /* backupBranchLayout */ true));
+      }
+      // This method is invoked on UI dispatch thread
+      @UIEffect
+      @Override
+      public void onSuccess() {
+        // Match below returns empty string just because must return something
```
Comment (PawelLipski) [inline frontendUiTableImpl/src/main/java/com/virtuslab/gitmachete/frontend/ui/impl/table/EnhancedGraphTable.java:244]: Hmm TBH `Match` feels a bit artificial here (also due to the reason stated in this comment)... anyway, maybe it's not a good idea to handle the failures in `onSuccess` callback? Can we use `onFailure` or a similar callback for that purpose?

File: frontendUiTableImpl/src/main/java/com/virtuslab/gitmachete/frontend/ui/impl/table/EnhancedGraphTable.java [hunk 6]
```diff
@@ -208,13 +209,74 @@ private void refreshModel(GitRepository gitRepository, List<String> skippedBranc
...
+      @Override
+      public void run(ProgressIndicator indicator) {
+        runResult = Try.of(() -> RuntimeBinding.instantiateSoleImplementingClass(IGitMacheteRepositoryCache.class)
+            .getInstance(mainDirPath, gitDirPath).discoverLayoutAndCreateSnapshot())
+            .filterTry(repositorySnapshot -> repositorySnapshot.getRootBranches().size() > 0,
+                __ -> new EmptyRepositoryException())
+            .map(repositorySnapshot -> {
+              var branchLayoutWriter = project.getService(BranchLayoutWriterProvider.class).getBranchLayoutWriter();
+              var branchLayout = repositorySnapshot.getBranchLayout().getOrNull();
+              return Tuple(branchLayoutWriter, branchLayout, repositorySnapshot);
```
Comment (PawelLipski) [inline frontendUiTableImpl/src/main/java/com/virtuslab/gitmachete/frontend/ui/impl/table/EnhancedGraphTable.java:235]: I'd rather create a dedicated private static class for the result... tuples with 3+ fields of different types are confusing

File: frontendUiTableImpl/src/main/java/com/virtuslab/gitmachete/frontend/ui/impl/table/EnhancedGraphTable.java [hunk 7]
```diff
@@ -208,13 +209,74 @@ private void refreshModel(GitRepository gitRepository, List<String> skippedBranc
...
+    var gitDirPath = GitVfsUtils.getGitDirectoryPath(selectedRepository).toAbsolutePath();
+
+    @SuppressWarnings({"nullness", "guieffect"}) // To allow runResult be uninitialized and Tuple accept @Nullable IBranchLayout
+    var task = new Task.Backgroundable(project,
+        getString("string.GitMachete.EnhancedGraphTable.automatic-discover.task-title")) {
+      private Try<Tuple3<IBranchLayoutWriter, @Nullable IBranchLayout, IGitMacheteRepositorySnapshot>> runResult;
+
+      @Override
+      public void run(ProgressIndicator indicator) {
+        runResult = Try.of(() -> RuntimeBinding.instantiateSoleImplementingClass(IGitMacheteRepositoryCache.class)
```
Comment (PawelLipski) [inline frontendUiTableImpl/src/main/java/com/virtuslab/gitmachete/frontend/ui/impl/table/EnhancedGraphTable.java:228]: TBH I think this code might read easier if un-FP-ified ;D Possibly even the failure/success can be dispatched onto UI thread directly from this method (with no need of `runResult`)... just musing over ;)

File: frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/speedsearch/DiscoverAction.java [hunk 8]
```diff
@@ -38,51 +39,69 @@ public IEnhancedLambdaLogger log() {
   @UIEffect
   public void actionPerformed(AnActionEvent anActionEvent) {
```
Comment (PawelLipski) [inline frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/speedsearch/DiscoverAction.java:40]: This action is going to be rarely tested in everyday use... consider adding a test case for discover to `UITestSuite` (as a first thing the test would need to wipe out or delete machete file)

File: frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/speedsearch/DiscoverAction.java (hunk 1 above)
Comment (micpiotrowski) [reply frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/speedsearch/DiscoverAction.java:1]: TBH, speed search is probably only place where this action is available, so this is the reason of this change.
But ok, `unanchored` is also ok.

File: frontendUiTableImpl/src/main/java/com/virtuslab/gitmachete/frontend/ui/impl/table/EnhancedGraphTable.java (hunk 5 above)
Comment (micpiotrowski) [reply frontendUiTableImpl/src/main/java/com/virtuslab/gitmachete/frontend/ui/impl/table/EnhancedGraphTable.java:244]: Unfortunately with `Task.Backgroundable` not. I try use a `Task.WithResult` but handling exception when receive the result is awkward (this is how it works in that case).

File: frontendUiTableImpl/src/main/java/com/virtuslab/gitmachete/frontend/ui/impl/table/EnhancedGraphTable.java (hunk 4 above)
Comment (micpiotrowski) [reply frontendUiTableImpl/src/main/java/com/virtuslab/gitmachete/frontend/ui/impl/table/EnhancedGraphTable.java:221]: No, this is about this field is uninitialized by constructor: `error: [[fbc, nullness, allcheckers]:initialization.fields.uninitialized] the constructor does not initialize fields: runResult`

Comment (PawelLipski) [reply frontendUiTableImpl/src/main/java/com/virtuslab/gitmachete/frontend/ui/impl/table/EnhancedGraphTable.java:221]: Ahh right... then maybe mark this field as `@MonotonicNonNull`?

File: frontendUiTableImpl/src/main/java/com/virtuslab/gitmachete/frontend/ui/impl/table/EnhancedGraphTable.java [hunk 9]
```diff
@@ -276,6 +272,12 @@ public void onSuccess() {
 
   private static final class EmptyRepositoryException extends Exception {}
   private static final class EmptyBranchLayoutException extends Exception {}
+  @RequiredArgsConstructor(staticName = "of")
```
Comment (PawelLipski) [inline frontendUiTableImpl/src/main/java/com/virtuslab/gitmachete/frontend/ui/impl/table/EnhancedGraphTable.java:275]: Nit: add empty line before this one

File: frontendUiTableImpl/src/main/java/com/virtuslab/gitmachete/frontend/ui/impl/table/EnhancedGraphTable.java [hunk 10]
```diff
@@ -218,51 +216,49 @@ private void doAutomaticDiscover(Path macheteFilePath) {
...
         getString("string.GitMachete.EnhancedGraphTable.automatic-discover.task-title")) {
-      private Try<Tuple3<IBranchLayoutWriter, @Nullable IBranchLayout, IGitMacheteRepositorySnapshot>> runResult;
+      private Try<DiscoverResult> discoverResult;
 
       @Override
       public void run(ProgressIndicator indicator) {
-        runResult = Try.of(() -> RuntimeBinding.instantiateSoleImplementingClass(IGitMacheteRepositoryCache.class)
+        discoverResult = Try.of(() -> RuntimeBinding.instantiateSoleImplementingClass(IGitMacheteRepositoryCache.class)
             .getInstance(mainDirPath, gitDirPath).discoverLayoutAndCreateSnapshot())
             .filterTry(repositorySnapshot -> repositorySnapshot.getRootBranches().size() > 0,
```
Comment (PawelLipski) [inline frontendUiTableImpl/src/main/java/com/virtuslab/gitmachete/frontend/ui/impl/table/EnhancedGraphTable.java:229]: Hmm how about the following: in case the condition is not satisfied, just `GuiUtils.invokeLater` the `setTextForEmptyTable(...)` piece and return from the method? Similarly for the case where `result.branchLayout == null` below. This way, the intermediate exceptions won't be needed and the entire flow will be likely easier to analyze

File: frontendUiTableImpl/src/main/java/com/virtuslab/gitmachete/frontend/ui/impl/table/EnhancedGraphTable.java [hunk 11]
```diff
@@ -208,11 +207,76 @@ private void refreshModel(GitRepository gitRepository, List<String> skippedBranc
...
+            .filterTry(result -> result.branchLayout != null, __ -> new EmptyBranchLayoutException())
+            .andThenTry(
+                result -> result.branchLayoutWriter.write(macheteFilePath, result.branchLayout, /* backupBranchLayout */ true));
+      }
+      // This method is invoked on UI dispatch thread
+      @UIEffect
+      @Override
+      public void onSuccess() {
+        discoverResult.onFailure(exception -> {
+          Class<? extends Throwable> exceptionClass = exception.getClass();
```
Comment (PawelLipski) [inline frontendUiTableImpl/src/main/java/com/virtuslab/gitmachete/frontend/ui/impl/table/EnhancedGraphTable.java:245]: Hmm no idea why this failing on `FenumChecker` :thinking: anyway, possibly as in the[comment](https://github.com/VirtusLab/git-machete-intellij-plugin/pull/550/files#r478652551) above, the temporary exception classes can be eliminated

File: frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/speedsearch/DiscoverAction.java (hunk 1 above)
Comment (mkondratek) [reply frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/speedsearch/DiscoverAction.java]: CMIIW but I dare timidly suppose that you misunderstand the meaning of `speed search`. 

This is `Find Action...` dialog (?):
image

This is speed search functionality:
image

No action is "available from `speed search`".

Comment (PawelLipski) [reply frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/speedsearch/DiscoverAction.java]: One can argue that speed search is also wired up to Ctrl+Shift+A... as in, IIUC it's "speed search" in the sense that it yields the results as you type (as opposed to how e.g. `grep` operates - only yielding the results once the search pattern is considered complete)

File: frontendUiTableImpl/src/main/java/com/virtuslab/gitmachete/frontend/ui/impl/table/EnhancedGraphTable.java (hunk 4 above)
Comment (micpiotrowski) [reply frontendUiTableImpl/src/main/java/com/virtuslab/gitmachete/frontend/ui/impl/table/EnhancedGraphTable.java]: Get rid of this field completely

File: frontendUiTableImpl/src/main/java/com/virtuslab/gitmachete/frontend/ui/impl/table/EnhancedGraphTable.java (hunk 5 above)
Comment (micpiotrowski) [reply frontendUiTableImpl/src/main/java/com/virtuslab/gitmachete/frontend/ui/impl/table/EnhancedGraphTable.java]: Out of date. Whole job is done in `run` method.

File: frontendUiTableImpl/src/main/java/com/virtuslab/gitmachete/frontend/ui/impl/table/EnhancedGraphTable.java (hunk 7 above)
Comment (micpiotrowski) [reply frontendUiTableImpl/src/main/java/com/virtuslab/gitmachete/frontend/ui/impl/table/EnhancedGraphTable.java]: revamped

File: frontendUiTableImpl/src/main/java/com/virtuslab/gitmachete/frontend/ui/impl/table/EnhancedGraphTable.java (hunk 10 above)
Comment (micpiotrowski) [reply frontendUiTableImpl/src/main/java/com/virtuslab/gitmachete/frontend/ui/impl/table/EnhancedGraphTable.java:229]: revamped

File: frontendUiTableImpl/src/main/java/com/virtuslab/gitmachete/frontend/ui/impl/table/EnhancedGraphTable.java (hunk 11 above)
Comment (micpiotrowski) [reply frontendUiTableImpl/src/main/java/com/virtuslab/gitmachete/frontend/ui/impl/table/EnhancedGraphTable.java:245]: get rid of this

File: frontendUiTableImpl/src/main/java/com/virtuslab/gitmachete/frontend/ui/impl/table/EnhancedGraphTable.java [hunk 12]
```diff
@@ -216,67 +213,57 @@ private void doAutomaticDiscover(Path macheteFilePath) {
...
-                exception.getMessage() != null ? exception.getMessage() : "");
-          }
-        })
-            .onSuccess(result -> {
-              VcsNotifier.getInstance(project)
-                  .notifyInfo(getString("string.GitMachete.EnhancedGraphTable.automatic-discover.success-message"));
-
-              gitMacheteRepositorySnapshot = result.repositorySnapshot;
+                exception.getMessage() != null ? exception.getMessage() : ""), NON_MODAL))
+            .onSuccess(v -> {
```
Comment (PawelLipski) [inline frontendUiTableImpl/src/main/java/com/virtuslab/gitmachete/frontend/ui/impl/table/EnhancedGraphTable.java:253]: Nit:
```suggestion
            .onSuccess(__ -> {
```

File: frontendUiTableImpl/src/main/java/com/virtuslab/gitmachete/frontend/ui/impl/table/EnhancedGraphTable.java [hunk 13]
```diff
@@ -216,67 +213,57 @@ private void doAutomaticDiscover(Path macheteFilePath) {
...
+
+        var branchLayoutWriter = project.getService(BranchLayoutWriterProvider.class).getBranchLayoutWriter();
+        var branchLayout = repositorySnapshot.getBranchLayout().getOrNull();
+
+        if (branchLayout == null) {
+          LOG.error("Can't get branch layout from repository snapshot");
+          return;
+        }
+
+        @SuppressWarnings("guieffect") // To allow lambda invocation in `onSuccess` method
```
Comment (PawelLipski) [inline frontendUiTableImpl/src/main/java/com/virtuslab/gitmachete/frontend/ui/impl/table/EnhancedGraphTable.java:248]: Hmm I'm afraid that the GuiEffect error here is justified :thinking:
See the comment below to `setModel`

File: frontendUiTableImpl/src/main/java/com/virtuslab/gitmachete/frontend/ui/impl/table/EnhancedGraphTable.java [hunk 14]
```diff
@@ -216,67 +213,57 @@ private void doAutomaticDiscover(Path macheteFilePath) {
...
+            .onSuccess(v -> {
+              GuiUtils.invokeLaterIfNeeded(
+                  () -> VcsNotifier.getInstance(project)
+                      .notifyInfo(getString("string.GitMachete.EnhancedGraphTable.automatic-discover.success-message")),
+                  NON_MODAL);
+
+              gitMacheteRepositorySnapshot = repositorySnapshot;
               var repositoryGraph = repositoryGraphCache.getRepositoryGraph(gitMacheteRepositorySnapshot, isListingCommits);
               setModel(new GraphTableModel(repositoryGraph));
               repaint();
```
Comment (PawelLipski) [inline frontendUiTableImpl/src/main/java/com/virtuslab/gitmachete/frontend/ui/impl/table/EnhancedGraphTable.java]: Hmm shouldn't those lines be executed on UI thread? :thinking:

File: frontendUiTableImpl/src/main/java/com/virtuslab/gitmachete/frontend/ui/impl/table/EnhancedGraphTable.java (hunk 12 above)
Comment (micpiotrowski) [reply frontendUiTableImpl/src/main/java/com/virtuslab/gitmachete/frontend/ui/impl/table/EnhancedGraphTable.java:253]: Probably should be even `___` (triple `_`) coz double `_` is already used in this scope

File: src/main/resources/META-INF/plugin.xml [hunk 15]
```diff
@@ -199,7 +199,7 @@
             from `Search Everywhere` and by link displayed in graph table when `machete` file
             is empty or absent -->
         <action id="GitMachete.DiscoverAction"
-                class="com.virtuslab.gitmachete.frontend.actions.graphtable.DiscoverAction"
+                class="com.virtuslab.gitmachete.frontend.actions.unanchored.DiscoverAction"
```
Comment (PawelLipski) [inline src/main/resources/META-INF/plugin.xml:202]: One file to update is also docs/plugins.jetbrains.com/general_usage_instructions.html - there's still a TODO for #176 there :/

File: frontendUiTableImpl/src/main/java/com/virtuslab/gitmachete/frontend/ui/impl/table/EnhancedGraphTable.java [hunk 16]
```diff
@@ -176,26 +180,18 @@ private void refreshModel(GitRepository gitRepository, List<String> skippedBranc
     } else {
       repositoryGraph = repositoryGraphCache.getRepositoryGraph(gitMacheteRepositorySnapshot, isListingCommits);
       if (gitMacheteRepositorySnapshot.getRootBranches().isEmpty()) {
-        setTextForEmptyTable(
```
Comment (PawelLipski) [inline frontendUiTableImpl/src/main/java/com/virtuslab/gitmachete/frontend/ui/impl/table/EnhancedGraphTable.java:179]: Line 162 still has a no longer relevant TODO for #176

**Comment by PawelLipski** [comment]
Taking over from @micpiotrowski...

File: frontendUiTableImpl/src/main/java/com/virtuslab/gitmachete/frontend/ui/impl/table/EnhancedGraphTable.java (hunk 13 above)
Comment (PawelLipski) [reply frontendUiTableImpl/src/main/java/com/virtuslab/gitmachete/frontend/ui/impl/table/EnhancedGraphTable.java:248]: Done, reworked

File: frontendUiTableImpl/src/main/java/com/virtuslab/gitmachete/frontend/ui/impl/table/EnhancedGraphTable.java (hunk 12 above)
Comment (PawelLipski) [reply frontendUiTableImpl/src/main/java/com/virtuslab/gitmachete/frontend/ui/impl/table/EnhancedGraphTable.java:253]: Reworked, no longer relevant

File: frontendUiTableImpl/src/main/java/com/virtuslab/gitmachete/frontend/ui/impl/table/EnhancedGraphTable.java (hunk 14 above)
Comment (PawelLipski) [reply frontendUiTableImpl/src/main/java/com/virtuslab/gitmachete/frontend/ui/impl/table/EnhancedGraphTable.java]: Done

**Comment by mkondratek** [comment]
I am not sure if this is an issue but consider the following scenario:

1. Machete file has a valid content (branches exists, properly indented)
2. Remove the whole content
3. Insert `secret` (assuming there is no branch called `secret`) in the first line
4. Refresh table (reload machete changes)

Actual:
`secret` "lost", replaced by the discovered layout

Expected:
A notification saying that `secret` does not belong to the local repo???

**Comment by mkondratek** [comment]
Another case is when there is no local branch withing the repo. Shouldn't refresh show some notification (`"No local branches within the repo"` or smth)? Presently nothing happens

**Comment by PawelLipski** [comment]
> Another case is when there is no local branch withing the repo. Shouldn't refresh show some notification (`"No local branches within the repo"` or smth)? Presently nothing happens

#562 opened

**Review by mkondratek — APPROVED** [review]
gj

**Review by PawelLipski — APPROVED** [review]
